python/ray/experimental/data/dataset.py

```diff
@@ -219,7 +219,41 @@ def split(self, n: int,
         Returns:
             A list of ``n`` disjoint dataset splits.
         """
-        raise NotImplementedError  # P1
+        assert n > 0 and n <= len(self._blocks)
+        assert locality_hints is None or len(locality_hints) == n
+
+        # split self._blocks into n chunks. the first {num_large_chunks} chunks
+        # have {small_chunk_size + 1} blocks, and remaining
+        # chunks have {small_chunk_size} blocks.
+        small_chunk_size = len(self._blocks) // n
+        num_large_chunks = len(self._blocks) - small_chunk_size * n
+
+        datasets = []
+        cur = 0
+        for i in range(n):
+            end = cur + small_chunk_size
+            if i < num_large_chunks:
+                end = end + 1
+            datasets.append(Dataset(self._blocks.slice(cur, end)))
+            cur = end
+
+        if locality_hints is None:
+            return datasets
+
+        # placement hints are implemented as a dummy map transform
+        # on the dataset with enforcing the ip_address resource equals to
+        # hinting actor's address.
+        actors_state = ray.state.actors()
+        for i, dataset in enumerate(datasets):
+            ip_address = actors_state.get(locality_hints[i]._actor_id.hex(),
+                                          {}).get("Address",
+                                                  {}).get("IPAddress")
+            if not ip_address:
+                # since it's best effort, skip if ip_address is None
+                continue
+            datasets[i] = dataset.map(
+                lambda x: x, resources={f"node:{ip_address}": 0.0001})
```

Comment:
We shouldn't move the data (that will happen anyways on read...). It's sufficient to preferentially assign blocks already colocated with the actor to that actors split. This can be done by retrieving the block's current location and matching it with the actor locations.